zerver/migrations/0023_subscription_flags.py

```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+import bitfield.models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('zerver', '0022_subscription_pin_to_top'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='subscription',
+            name='flags',
+            field=bitfield.models.BitField([b'stream_admin', b'moderator', b'regular', b'read_only'], default=0),
```

Comment:
Yeah, I agree, we should do this based on permissions in the style you describe @neerajwahi, rather than roles, at least one the backend, since we'll probably eventually want to enable users to define sets of permissions that don't form a single hierarchy.  

One thing that this does point out to me is that we'll probably need a "default user permissions" field to store what permissions users who subscribe to the stream (or are subscribed by another user) get on joining.